File: TriCoachAI/Core/PlanEngine/RuleBasedPlanStrategy.swift
import Foundation

/// Default plan generation strategy: deterministic periodization rules
/// (no network, no LLM). See `PlanGenerationStrategy` for the extensibility
/// point a future `LLMPlanStrategy` would implement instead.
struct RuleBasedPlanStrategy: PlanGenerationStrategy {
    func generate(context: PlanGenerationContext) -> TrainingPlan {
        let totalWeeks = min(max(context.explicitDurationWeeks ?? context.goal.weeksUntilTarget, 4), 52)
        let calendar = Calendar.current
        let startDate = calendar.startOfDay(for: context.startDate)
        let endDate = calendar.date(byAdding: .day, value: totalWeeks * 7 - 1, to: startDate) ?? startDate

        var macrocycles = PeriodizationEngine.buildMacrocycles(
            totalWeeks: totalWeeks,
            goalType: context.goal.type,
            startDate: startDate
        )

        var globalWeekIndex = 1
        for macroIndex in macrocycles.indices {
            var mesocycles = MesocycleBuilder.build(macrocycle: macrocycles[macroIndex])
            for mesoIndex in mesocycles.indices {
                let microcycles = MicrocycleBuilder.build(
                    mesocycle: mesocycles[mesoIndex],
                    phase: macrocycles[macroIndex].phase,
                    startingGlobalWeekIndex: globalWeekIndex,
                    totalPlanWeeks: totalWeeks,
                    context: context
                )
                mesocycles[mesoIndex].microcycles = microcycles
                globalWeekIndex += microcycles.count
            }
            macrocycles[macroIndex].mesocycles = mesocycles
        }

        return TrainingPlan(
            goalId: context.goal.id,
            startDate: startDate,
            endDate: endDate,
            durationWeeks: totalWeeks,
            status: .active,
            macrocycles: macrocycles
        )
    }
}

File: tricoach-backend/src/modules/plans/engine/ruleBasedPlanStrategy.ts
import { randomUUID } from 'crypto';
import { MesocycleBuilder } from './mesocycleBuilder';
import { MicrocycleBuilder } from './microcycleBuilder';
import { addDays, PeriodizationEngine } from './periodizationEngine';
import { PlanGenerationContext, TrainingPlan } from './types';
import { weeksUntilTarget } from './goalTypeInfo';

/** Direct port of Core/PlanEngine/RuleBasedPlanStrategy.swift. */
export function generatePlan(context: PlanGenerationContext): TrainingPlan {
  const totalWeeks = Math.min(
    Math.max(context.explicitDurationWeeks ?? weeksUntilTarget(context.goal.targetDate, context.goal.type, context.startDate), 4),
    52
  );
  const startDate = new Date(context.startDate);
  startDate.setHours(0, 0, 0, 0);
  const endDate = addDays(startDate, totalWeeks * 7 - 1);

  const macrocycles = PeriodizationEngine.buildMacrocycles(totalWeeks, context.goal.type, startDate);

  let globalWeekIndex = 1;
  for (const macrocycle of macrocycles) {
    const mesocycles = MesocycleBuilder.build(macrocycle);
    for (const mesocycle of mesocycles) {
      const microcycles = MicrocycleBuilder.build(mesocycle, macrocycle.phase, globalWeekIndex, totalWeeks, context);
      mesocycle.microcycles = microcycles;
      globalWeekIndex += microcycles.length;
    }
    macrocycle.mesocycles = mesocycles;
  }

  return {
    id: randomUUID(),
    goalId: context.goal.id,
    startDate,
    endDate,
    durationWeeks: totalWeeks,
    status: 'active',
    generationVersion: 'rule-based-v1',
    createdAt: new Date(),
    macrocycles,
  };
}
